### ISSUE TITLE
"Relative import paths need explicit file extensions in ECMAScript imports" errors when consuming ESM side of package

### ISSUE THREAD

**Issue by Nikovic** [planttheidea/fast-equals#124]
Greetings!

First off, thank you for creating and maintaining fast-equals.

I'm trying to use it in my own TypeScript lib/package, but when building (the CJS side of my lib) with tsc, I get
`error TS2834: Relative import paths need explicit file extensions in ECMAScript imports when '--moduleResolution' is 'node16' or 'nodenext'. Consider adding an extension to the import path.`

I only recently found out myself that node/tsc "don't like" omitted file extensions. Instead one should (mostly?) use '.js', even when importing from (.d).ts files (i.e. types and all). See [this TS documenation page](https://www.typescriptlang.org/docs/handbook/modules/reference.html#the-moduleresolution-compiler-option), where the lookup rules are detailed.

I got my package to build locally by adjusting the few imports in `node_modules/fast-equals/dist/esm/types/index.d.ts` and `node_modules/fast-equals/dist/esm/types/utils.d.ts`, but I assume if I were to publish like that, the next person down the code-use/compilation chain would be greeted with the same error.

So I'd like to ask you to check if adding the '.js' to imports on the source side of things is feasible (i.e. doesn't break things elsewhere). It _could_ be a quick and easy thing to fix (or a configuration nightmare with no end. I'd say the chances are 50:50 😂).

Thank you, and have a great rest of 12024 (Human Era calendar).

**Comment by planttheidea** [comment]
I think I can try to approach this in the coming days, but because it requires a tweak to the distributed file I'll have to do it as a separate release. I just tried playing with it locally, and I can replicate the issue and have a rudimentary version working, but it requires more rigorous checks for consumption. It requires a tweak to the TS configuration, but may require a custom transform on the output types to change `.ts` to `.d.ts`.

Expect an update early next week.

**Comment by planttheidea** [comment]
Alright so I _should_ be able to just use `.js` extension as you astutely mentioned, and it avoids the seemingly-conflicting use of `.ts`, which requires the `allowImportingTsExtensions` to be active. That said, considering the nature of the change, I'd like to test it independently.

Basically, I'll publish a beta (installable via `npm install fast-equals@next`) with the `.js` extensions. If you can install and consume the package without issue, then I can publish it to the official version. Let me know if you're willing to validate the change.

**Comment by Nikovic** [comment]
Yes, I can test this later today/whenever the beta is ready. Thanks again.

**Comment by planttheidea** [comment]
Alright this turned out to be trickier than expected; while there was no TS issue with applying the `.js` file extensions, the build system as well as `jest` were complaining about the specific extensions that were not valid. So, I went more of a hack direction, which was to apply the file extension in the type files only as a script that runs post-build.

I've published `5.2.2-beta.0` with these changes; :crossed_fingers: it does the thing.

**Comment by Nikovic** [comment]
I can confirm that importing and using 5.2.2-beta.0 works!
(Given my setup, which doesn't do anything unusual afaict)

Thank you!

**Comment by planttheidea** [comment]
Published officially as `5.2.2`. Thanks for confirming!
